README
NAME

    HTML::Formatter - Base class for HTML formatters

VERSION

    version 2.15

SYNOPSIS

      use HTML::FormatSomething;
      my $infile  = "whatever.html";
      my $outfile = "whatever.file";
      open OUT, ">$outfile"
       or die "Can't write-open $outfile: $!\n";
    
      print OUT HTML::FormatSomething->format_file(
        $infile,
          'option1' => 'value1',
          'option2' => 'value2',
          ...
      );
      close(OUT);

DESCRIPTION

    HTML::Formatter is a base class for classes that take HTML and format
    it to some output format. When you take an object of such a base class
    and call $formatter-format( $tree )> with an HTML::TreeBuilder (or
    HTML::Element) object, they return the appropriately formatted string
    for the input HTML.

    HTML formatters are able to format a HTML syntax tree into various
    printable formats. Different formatters produce output for different
    output media. Common for all formatters are that they will return the
    formatted output when the format() method is called. The format()
    method takes a HTML::Element object (usually the HTML::TreeBuilder root
    object) as parameter.

    The distribution name has been changed to HTML-Formatter as detailed in
    "DISTRIBUTION NAME"

METHODS

 new

        my $formatter = FormatterClass->new(
            option1 => value1, option2 => value2, ...
        );

    This creates a new formatter object with the given options.

 format_file

 format_from_file

        $string = FormatterClass->format_file(
            $html_source,
            option1 => value1, option2 => value2, ...
            );

    Return a string consisting of the result of using the given class to
    format the given HTML file according to the given (optional) options.
    Internally it calls SomeClass->new( ... )->format( ... ) on a new
    HTML::TreeBuilder object based on the given HTML file.

 format_string

 format_from_string

        $string = FormatterClass->format_string(
            $html_source,
            option1 => value1, option2 => value2, ...
            );

    Return a string consisting of the result of using the given class to
    format the given HTML source according to the given (optional) options.
    Internally it calls SomeClass->new( ... )->format( ... ) on a new
    HTML::TreeBuilder object based on the given source.

 format

        my $render_string = $formatter->format( $html_tree_object );

    This renders the given HTML object according to the options set for
    $formatter.

    After you've used a particular formatter object to format a particular
    HTML tree object, you probably should not use either again.

DISTRIBUTION NAME

    This module was originally named HTML-Format despite not containing a
    HTML::Format module within it. As rules on naming have been taken more
    seriously, and the PAUSE <https://pause.perl.org/> toolchain adapted so
    that getting the distribution indexed was more difficult, it became
    obvious that I should rename the distribution to HTML-Formatter
    matching the base HTML::Formatter module.

    As of release 2.13 this is released as the HTML-Formatter distribution
    with corresponding changes to the git repository name and associated
    items.

    Due to the way that the module is put together this should have no
    effect on code using the module. The only issues will be where the
    distribution name was used within dependancies.

SEE ALSO

    The three specific formatters:-

    HTML::FormatText

      Format HTML into plain text

    HTML::FormatPS

      Format HTML into postscript

    HTML::FormatRTF

      Format HTML into Rich Text Format

    Also the HTML manipulation libraries used - HTML::TreeBuilder,
    HTML::Element and HTML::Tree

INSTALLATION

    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS

    You can make new bug reports, and view existing ones, through the web
    interface at
    http://rt.cpan.org/Public/Dist/Display.html?Name=HTML-Formatter.

AVAILABILITY

    The project homepage is https://metacpan.org/release/HTML-Formatter.

    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit http://www.perl.com/CPAN/ to find a
    CPAN site near you, or see
    https://metacpan.org/module/HTML::Formatter/.

AUTHORS

      * Nigel Metheringham <nigelm@cpan.org>

      * Sean M Burke <sburke@cpan.org>

      * Gisle Aas <gisle@ActiveState.com>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2016 by Nigel Metheringham, 2002-2005
    Sean M Burke, 1999-2002 Gisle Aas.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
{{$NEXT}}

2.16      2016-12-07 09:12:41+00:00 Europe/London
    - Public release after a period on test with the smokers

2.15      2016-12-01 15:40:16+00:00 Europe/London (TRIAL RELEASE)
    - Deal with some relatively harmless test failures
    - RT#11783 fixes for Markdown fragment issues
    - Series of changes for UTF support from Sam Kington
    - Better travis handling

2.14      2015-10-13 14:46:01+01:00 Europe/London
    - Added Travis CI status to main documentation
    - First official release under new distribution name following trial release.

2.13      2015-10-12 16:26:46+01:00 Europe/London (TRIAL RELEASE)
    - Renamed distribution from HTML-Format -> HTML-Formatter
    - Documentation update within the main POD to reflect name change
    - No functional changes

2.12      2015-10-10 17:49:45+01:00 Europe/London
    - Minor test related fixes
    - Transition to using File::Slurper in place of File::Slurp
      Thanks to Karen Etheridge for the patch
    - Add Travis CI integration

2.11      2013-10-27 20:57:23 Europe/London
    - Typo fixes from dsteinbrunner
      https://github.com/nigelm/html-format/pull/2
    - Markdown support from mrallen1
      https://github.com/nigelm/html-format/pull/1

2.10      2011-07-18 17:07:35 Europe/London
    - Tests reworked to not use subtests.  Development versions
      of Test::More do not support subtests, and the smokers
      that are running with these dev versions then send me
      confusing test failures.  This is a pain.

2.09      2011-07-15 14:20:12 Europe/London
    - Reworking of HTML quote issue in FormatPS
    - Now remap all double quotes to " in FormatPS (which is not
      really right, but the best I can do with latin1 output)

2.08      2011-07-13 16:20:54 Europe/London
    - RT#69426 - issues with HTML quotes
    - UTF8 handling requires perl 5.8, so minimum requirement bumped

2.07      2011-04-20 16:08:15 Europe/London
    - Release unchanged except for version number after a period
      to see if test release threw up any issues...

2.06      2011-04-15 12:01:33 Europe/London
    - Code gardening - no functionality modifications - includes:-
    - Moved old DEBUG framework to Smart::Comments
    - Ensured variables are declared my as much as possible
    - Ensure strict/warnings used as far as possible
    - use base rather than playing games with @ISA
    - Changes above enforce a requirement of perl 5.6.1
    - Robo-reformatted by perltidy
    - Stripped function prototypes where used
    - Passes standard perl critic with one exclusion
    - Tweaks for comprehensibility - splitting sections etc

2.05      2011-02-18 11:28:33 Europe/London
    - Smoke tests on windows showed needed make PS tests less strict
    - Changed the github location to be all lowercase
    - Releasing this as a stop gap - will do further work now
      incompatibilities dealt with

2.04.02   2011-02-16 11:30:24 Europe/London
    - Added dependancies for font metrics files
    - Documentation tidy up - but needs further work
    - new TRIAL release - intend to push this as full release if OK

2.04.01   2011-02-14 20:26:07 Europe/London
    - Git conversion
    - Dist::Zilla conversion
    - RT#54636 - UNIVERSAL::Can deprecated
    - RT#56278 - RTF formatting parameters ignored by new
    - Start of a new test harness
    - TRIAL release pushed out.  Aiming to improve docs and maybe
      hit rest of RT queue before a full release.

2.04      2004-06-01   Sean M. Burke <sburke@cpan.org>
    - just cosmetic doc changes
    - Also, bumping all the version numbers up to 2.04 just for fun.


2.03      2002-11-22   Sean M. Burke <sburke@cpan.org>
    - minor bugfix
      The lack of \plain's in HTML::FormatRTF's output was confusing some
      formatters.  Fixed.


2.02      2002-11-16   Sean M. Burke <sburke@cpan.org>
    - minor bugfix
      Changed the set_version_tag method in HTML::Formatter to avoid some
      undef warnings that "James W. Durkin" <jwd@phonogram.net> found.

      No big bug reports on the previous version, so maybe it's not so
      experimental after all!


2.01      2002-11-07   Sean M. Burke <sburke@cpan.org>
    - experimental major version change
    - Added HTML::FormatRTF.
    - Lots of internal changes to HTML::FormatPS and HTML::Formatter.
    - Added format_file and format_string methods to HTML::Formatter.
    - Lots of little tidying up in the docs.


1.24      2002-10-29   Sean M. Burke <sburke@cpan.org>
    - Just taking it over from Gisle, and releasing this to make sure
      CPAN et al register this change.


1.23      2000-06-09   Gisle Aas <gisle@ActiveState.com>
    - Skip <script> and <style> stuff.


1.22      1999-12-16   Gisle Aas <gisle@aas.no>
    - Release HTML-Format-1.22
    - Splitted off HTML::Format* modules from HTML-Tree when Sean M. Burke
      took over maintainance.  You should have HTML-Tree version 0.61 or
      better installed before installing this.
    - Added some simple formatter tests.


0.53      1999-12-15   Gisle Aas <gisle@aas.no>
    - Release HTML-Tree-0.53
    - Make it compatible with HTML-Parser-3.00


0.52      1999-11-10   Gisle Aas <gisle@aas.no>
    - Fix SYNOPSIS for HTML::FormatText as suggested by
      Michael G Schwern <schwern@pobox.com>
    - Updated my email address.


0.51      1998-07-07   Gisle Aas <aas@sn.no>
    - Avoid new warnings introduced by perl5.004_70


0.50 1998-04-01   Gisle Aas <aas@sn.no>
    - The HTML::* modules the dealt with HTML syntax trees
      was unbundled from libwww-perl-5.22.



t/lib/Test/HTML/Formatter.pm
package Test::HTML::Formatter;

use strict;
use warnings;
use Test::More;

sub test_files {
    my ($test_class, %args) = @_;

    # Pull in our class.
    my $class = 'HTML::' . $args{class_suffix};
    use_ok($class);

    # Find the files we want to test.
    foreach my $infile ( glob( File::Spec->catfile( 't', 'data', 'in', '*.html' ) ) ) {
        my $obj = new_ok($class);
        ok( -f $infile, "Testing file handling for $infile" );
        my $expfilename = ( File::Spec->splitpath($infile) )[2];
        $expfilename =~ s/\.html$/.$args{filename_extension}/i;
        my $expfile = File::Spec->catfile( 't', 'data', 'expected', $expfilename );
        ok( -f $expfile, "  Expected result file $expfile exists" );
        if (-f $expfile) {
            SKIP:
            {
                if (-s $expfile == 0) {
                    skip "No idea what $expfile should look like", 1;
                }
                $args{callback_test_file}->($obj, $infile, $expfile);
            }
        }
    }
}

1;

t/support/generate_results.pl
#!/usr/bin/env perl
#
# Generate the output test files
#
# Obviously just building these in the same way as we do the tests
# means they will always pass - so you should be checking these
# carefully!
#
use strict;
use warnings;
use File::Spec;    # try to keep pathnames neutral
use File::Slurper 'write_binary';
use HTML::FormatRTF;
use HTML::FormatPS;
use HTML::FormatText;
use HTML::FormatMarkdown;

foreach my $infile ( glob( File::Spec->catfile( 't', 'data', 'in', '*.html' ) ) ) {
    my $outfile =
        substr( File::Spec->catfile( 't', 'data', 'expected', ( File::Spec->splitpath($infile) )[2] ), 0, -4 );
    write_binary( ( $outfile . 'ps' ), HTML::FormatPS->format_file( $infile, leftmargin => 5, rightmargin => 50 ) );
    write_binary( ( $outfile . 'rtf' ), HTML::FormatRTF->format_file( $infile, leftmargin => 5, rightmargin => 50 ) );
    write_binary( ( $outfile . 'txt' ), HTML::FormatText->format_file( $infile, leftmargin => 5, rightmargin => 50 ) );
    write_binary( ( $outfile . 'md' ), HTML::FormatMarkdown->format_file( $infile, leftmargin => 5, rightmargin => 50 ) );
}


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.039.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28",
    "Test::CPAN::Meta" => 0,
    "Test::EOL" => 0,
    "Test::NoTabs" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Base class for HTML formatters",
  "dist_author" => [
    "Nigel Metheringham <nigelm\@cpan.org>",
    "Sean M Burke <sburke\@cpan.org>",
    "Gisle Aas <gisle\@ActiveState.com>"
  ],
  "dist_name" => "HTML-Formatter",
  "dist_version" => "2.14",
  "license" => "perl",
  "module_name" => "HTML::Formatter",
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Encode" => 0,
    "Font::Metrics::Courier" => 0,
    "Font::Metrics::CourierBold" => 0,
    "Font::Metrics::CourierBoldOblique" => 0,
    "Font::Metrics::CourierOblique" => 0,
    "Font::Metrics::Helvetica" => 0,
    "Font::Metrics::HelveticaBold" => 0,
    "Font::Metrics::HelveticaBoldOblique" => 0,
    "Font::Metrics::HelveticaOblique" => 0,
    "Font::Metrics::TimesBold" => 0,
    "Font::Metrics::TimesBoldItalic" => 0,
    "Font::Metrics::TimesItalic" => 0,
    "Font::Metrics::TimesRoman" => 0,
    "HTML::Element" => "3.15",
    "HTML::TreeBuilder" => 0,
    "IO::File" => 0,
    "base" => 0,
    "integer" => 0,
    "parent" => 0,
    "perl" => "5.008",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "test_requires" => {
    "File::Slurper" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Scalar::Util" => 0,
    "Test::More" => "0.96",
    "blib" => "1.01",
    "bytes" => 0
  }
);


my %fallback_build_requires = (
  "File::Slurper" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.28",
  "Scalar::Util" => 0,
  "Test::CPAN::Meta" => 0,
  "Test::EOL" => 0,
  "Test::More" => "0.96",
  "Test::NoTabs" => 0,
  "blib" => "1.01",
  "bytes" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;


lib/HTML/FormatText.pm
package HTML::FormatText;

# ABSTRACT: Format HTML as plaintext


use 5.006_001;
use strict;
use warnings;

# We now use Smart::Comments in place of the old DEBUG framework.
# this should be commented out in release versions....
##use Smart::Comments;

use base 'HTML::Formatter';

our $VERSION = '2.15'; # TRIAL VERSION
our $AUTHORITY = 'cpan:NIGELM'; # AUTHORITY

# ------------------------------------------------------------------------
sub default_values {
    (   shift->SUPER::default_values(),
        lm => 3,     # left margin
        rm => 72,    # right margin (actually, maximum text width)
    );
}

# ------------------------------------------------------------------------
sub configure {
    my ( $self, $hash ) = @_;

    my $lm = $self->{lm};
    my $rm = $self->{rm};

    $lm = delete $hash->{lm}          if exists $hash->{lm};
    $lm = delete $hash->{leftmargin}  if exists $hash->{leftmargin};
    $rm = delete $hash->{rm}          if exists $hash->{rm};
    $rm = delete $hash->{rightmargin} if exists $hash->{rightmargin};

    my $width = $rm - $lm;
    if ( $width < 1 ) {
        warn "Bad margins, ignored" if $^W;
        return;
    }
    if ( $width < 20 ) {
        warn "Page probably too narrow" if $^W;
    }

    for ( keys %$hash ) {
        warn "Unknown configure option '$_'" if $^W;
    }

    $self->{lm} = $lm;
    $self->{rm} = $rm;
    $self;
}

# ------------------------------------------------------------------------
sub begin {
    my $self = shift;

    $self->SUPER::begin;
    $self->{curpos} = 0;    # current output position.
    $self->{maxpos} = 0;    # highest value of $pos (used by header underliner)
    $self->{hspace} = 0;    # horizontal space pending flag
}

# ------------------------------------------------------------------------
sub end {
    shift->collect("\n");
}

# ------------------------------------------------------------------------
sub header_start {
    my ( $self, $level ) = @_;

    $self->vspace( 1 + ( 6 - $level ) * 0.4 );
    $self->{maxpos} = 0;
    1;
}

# ------------------------------------------------------------------------
sub header_end {
    my ( $self, $level ) = @_;

    if ( $level <= 2 ) {
        my $line;
        $line = '=' if $level == 1;
        $line = '-' if $level == 2;
        $self->vspace(0);
        $self->out( $line x ( $self->{maxpos} - $self->{lm} ) );
    }
    $self->vspace(1);
    1;
}

# ------------------------------------------------------------------------
sub bullet {
    my $self = shift;

    $self->SUPER::bullet( $_[0] . ' ' );
}

# ------------------------------------------------------------------------
sub hr_start {
    my $self = shift;

    $self->vspace(1);
    $self->out( '-' x ( $self->{rm} - $self->{lm} ) );
    $self->vspace(1);
}

# ------------------------------------------------------------------------
sub pre_out {
    my $self = shift;

    # should really handle bold/italic etc.
    if ( defined $self->{vspace} ) {
        if ( $self->{out} ) {
            $self->nl() while $self->{vspace}-- >= 0;
            $self->{vspace} = undef;
        }
    }
    my $indent = ' ' x $self->{lm};
    my $pre    = shift;
    $pre =~ s/^/$indent/mg;
    $self->collect($pre);
    $self->{out}++;
}

# ------------------------------------------------------------------------
sub out {
    my $self = shift;
    my $text = shift;

    if ( $text =~ /^\s*$/ ) {
        $self->{hspace} = 1;
        return;
    }

    if ( defined $self->{vspace} ) {
        if ( $self->{out} ) {
            $self->nl while $self->{vspace}-- >= 0;
        }
        $self->goto_lm;
        $self->{vspace} = undef;
        $self->{hspace} = 0;
    }

    if ( $self->{hspace} ) {
        if ( $self->{curpos} + length($text) > $self->{rm} ) {

            # word will not fit on line; do a line break
            $self->nl;
            $self->goto_lm;
        }
        else {

            # word fits on line; use a space
            $self->collect(' ');
            ++$self->{curpos};
        }
        $self->{hspace} = 0;
    }

    $self->collect($text);
    my $pos = $self->{curpos} += length $text;
    $self->{maxpos} = $pos if $self->{maxpos} < $pos;
    $self->{'out'}++;
}

# ------------------------------------------------------------------------
sub goto_lm {
    my $self = shift;

    my $pos = $self->{curpos};
    my $lm  = $self->{lm};
    if ( $pos < $lm ) {
        $self->{curpos} = $lm;
        $self->collect( " " x ( $lm - $pos ) );
    }
}

# ------------------------------------------------------------------------
sub nl {
    my $self = shift;

    $self->{'out'}++;
    $self->{curpos} = 0;
    $self->collect("\n");
}

# ------------------------------------------------------------------------
sub adjust_lm {
    my $self = shift;

    $self->{lm} += $_[0];
    $self->goto_lm;
}

# ------------------------------------------------------------------------
sub adjust_rm {
    shift->{rm} += $_[0];
}


1;

__END__

=pod

=for test_synopsis 1;
__END__

=for stopwords latin1 leftmargin lm plaintext rightmargin rm CPAN homepage

=head1 NAME

HTML::FormatText - Format HTML as plaintext

=head1 VERSION

version 2.15

=head1 SYNOPSIS

    use HTML::TreeBuilder;
    $tree = HTML::TreeBuilder->new->parse_file("test.html");

    use HTML::FormatText;
    $formatter = HTML::FormatText->new(leftmargin => 0, rightmargin => 50);
    print $formatter->format($tree);

or, more simply:

    use HTML::FormatText;
    my $string = HTML::FormatText->format_file(
        'test.html',
        leftmargin => 0, rightmargin => 50
        );

=head1 DESCRIPTION

HTML::FormatText is a formatter that outputs plain text. All character
attributes (bold/italic/underline) are ignored. Formatting of HTML tables and
forms is not implemented.

HTML::FormatText is built on L<HTML::Formatter> and documentation for that
module applies to this - especially L<HTML::Formatter/new>,
L<HTML::Formatter/format_file> and L<HTML::Formatter/format_string>.

You might specify the following parameters when constructing the formatter:

=over 4

=item I<leftmargin> (alias I<lm>)

The column of the left margin. The default is 3.

=item I<rightmargin> (alias I<rm>)

The column of the right margin. The default is 72.

=back

=head1 SEE ALSO

L<HTML::Formatter>

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

You can make new bug reports, and view existing ones, through the
web interface at L<http://rt.cpan.org/Public/Dist/Display.html?Name=HTML-Formatter>.

=head1 AVAILABILITY

The project homepage is L<https://metacpan.org/release/HTML-Formatter>.

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see L<https://metacpan.org/module/HTML::Formatter/>.

=head1 AUTHORS

=over 4

=item *

Nigel Metheringham <nigelm@cpan.org>

=item *

Sean M Burke <sburke@cpan.org>

=item *

Gisle Aas <gisle@ActiveState.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Nigel Metheringham, 2002-2005 Sean M Burke, 1999-2002 Gisle Aas.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/HTML/FormatMarkdown.pm
package HTML::FormatMarkdown;

# ABSTRACT: Format HTML as Markdown


use 5.006_001;
use strict;
use warnings;

use parent 'HTML::Formatter';

our $VERSION = '2.15'; # TRIAL VERSION
our $AUTHORITY = 'cpan:NIGELM'; # AUTHORITY

sub default_values {
    (   shift->SUPER::default_values(),
        lm => 0,
        rm => 70,
    );
}

sub configure {
    my ( $self, $hash ) = @_;

    my $lm = $self->{lm};
    my $rm = $self->{rm};

    $lm = delete $hash->{lm}          if exists $hash->{lm};
    $lm = delete $hash->{leftmargin}  if exists $hash->{leftmargin};
    $rm = delete $hash->{rm}          if exists $hash->{rm};
    $rm = delete $hash->{rightmargin} if exists $hash->{rightmargin};

    my $width = $rm - $lm;
    if ( $width < 1 ) {
        warn "Bad margins, ignored" if $^W;
        return;
    }
    if ( $width < 20 ) {
        warn "Page probably too narrow" if $^W;
    }

    for ( keys %$hash ) {
        warn "Unknown configure option '$_'" if $^W;
    }

    $self->{lm} = $lm;
    $self->{rm} = $rm;
    $self;
}

sub begin {
    my $self = shift;

    $self->SUPER::begin();
    $self->{maxpos} = 0;
    $self->{curpos} = 0;    # current output position.
}

sub end {
    shift->collect("\n");
}

sub header_start {
    my ( $self, $level ) = @_;

    $self->vspace(1);
    $self->out( '#' x $level . ' ' );
    1;
}

sub header_end {
    my ( $self, $level ) = @_;

    $self->out( ' ' . '#' x $level );
    $self->vspace(1);
}

sub bullet {
    my $self = shift;

    $self->SUPER::bullet( $_[0] . ' ' );

}

sub hr_start {
    my $self = shift;

    $self->vspace(1);
    $self->out('- - -');
    $self->vspace(1);
}

sub img_start {
    my ( $self, $node ) = @_;

    my $alt = $node->attr('alt') || '';
    my $src = $node->attr('src');

    $self->out("![$alt]($src)");
}

sub a_start {
    my ( $self, $node ) = @_;

    # ignore named anchors
    if ( $node->attr('name') ) {
        1;
    }
    elsif ( $node->attr('href') =~ /^#/ ) {
        1;
    }
    else {
        $self->out("[");
        1;
    }
}

sub a_end {
    my ( $self, $node ) = @_;

    if ( $node->attr('name') ) {
        return;
    }
    elsif ( my $href = $node->attr('href') ) {
        if ( $href =~ /^#/ ) {
            return;
        }
        $self->out("]($href)");
    }
}

sub b_start { shift->out("**") }
sub b_end   { shift->out("**") }
sub i_start { shift->out("*") }
sub i_end   { shift->out("*") }

sub tt_start {
    my $self = shift;

    if ( $self->{pre} ) {
        return 1;
    }
    else {
        $self->out("`");
    }
}

sub tt_end {
    my $self = shift;

    if ( $self->{pre} ) {
        return;
    }
    else {
        $self->out("`");
    }
}

sub blockquote_start {
    my $self = shift;

    $self->{blockquote}++;
    $self->vspace(1);
    $self->adjust_rm(-4);

    1;
}

sub blockquote_end {
    my $self = shift;

    $self->{blockquote}--;
    $self->vspace(1);
    $self->adjust_rm(+4);

}

sub blockquote_out {
    my ( $self, $text ) = @_;

    $self->nl;
    $self->goto_lm;

    my $line = "> ";
    $self->{curpos} += 2;

    foreach my $word ( split /\s/, $text ) {
        $line .= "$word ";
        if ( ( $self->{curpos} + length($line) ) > $self->{rm} ) {
            $self->collect($line);
            $self->nl;
            $self->goto_lm;
            $line = "> ";
            $self->{curpos} += 2;
        }
    }

    $self->collect($line);
    $self->nl;

}

# Quoted from HTML::FormatText
sub pre_out {
    my $self = shift;

    if ( defined $self->{vspace} ) {
        if ( $self->{out} ) {
            $self->nl() while $self->{vspace}-- >= 0;
            $self->{vspace} = undef;
        }
    }

    my $indent = ' ' x $self->{lm};
    $indent .= ' ' x 4;
    my $pre = shift;
    $pre =~ s/^/$indent/mg;
    $self->collect($pre);
    $self->{out}++;
}

sub out {
    my $self = shift;
    my $text = shift;

    if ( $text =~ /^\s*$/ ) {
        $self->{hspace} = 1;
        return;
    }

    if ( defined $self->{vspace} ) {
        if ( $self->{out} ) {
            $self->nl while $self->{vspace}-- >= 0;
        }
        $self->goto_lm;
        $self->{vspace} = undef;
        $self->{hspace} = 0;
    }

    if ( $self->{hspace} ) {
        if ( $self->{curpos} + length($text) > $self->{rm} ) {

            # word will not fit on line; do a line break
            $self->nl;
            $self->goto_lm;
        }
        else {

            # word fits on line; use a space
            $self->collect(' ');
            ++$self->{curpos};
        }
        $self->{hspace} = 0;
    }

    $self->collect($text);
    my $pos = $self->{curpos} += length $text;
    $self->{maxpos} = $pos if $self->{maxpos} < $pos;
    $self->{'out'}++;
}

sub goto_lm {
    my $self = shift;

    my $pos = $self->{curpos};
    my $lm  = $self->{lm};
    if ( $pos < $lm ) {
        $self->{curpos} = $lm;
        $self->collect( " " x ( $lm - $pos ) );
    }
}

sub nl {
    my $self = shift;

    $self->{'out'}++;
    $self->{curpos} = 0;
    $self->collect("\n");
}

sub adjust_lm {
    my $self = shift;

    $self->{lm} += $_[0];
    $self->goto_lm;
}

sub adjust_rm {
    shift->{rm} += $_[0];
}

1;

__END__

=pod

=for stopwords CPAN Markdown homepage

=for test_synopsis 1;
__END__

=head1 NAME

HTML::FormatMarkdown - Format HTML as Markdown

=head1 VERSION

version 2.15

=head1 SYNOPSIS

    use HTML::FormatMarkdown;

    my $string = HTML::FormatMarkdown->format_file(
        'test.html'
    );

    open my $fh, ">", "test.md" or die "$!\n";
    print $fh $string;
    close $fh;

=head1 DESCRIPTION

HTML::FormatMarkdown is a formatter that outputs Markdown.

HTML::FormatMarkdown is built on L<HTML::Formatter> and documentation for that
module applies to this - especially L<HTML::Formatter/new>,
L<HTML::Formatter/format_file> and L<HTML::Formatter/format_string>.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

You can make new bug reports, and view existing ones, through the
web interface at L<http://rt.cpan.org/Public/Dist/Display.html?Name=HTML-Formatter>.

=head1 AVAILABILITY

The project homepage is L<https://metacpan.org/release/HTML-Formatter>.

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see L<https://metacpan.org/module/HTML::Formatter/>.

=head1 AUTHORS

=over 4

=item *

Nigel Metheringham <nigelm@cpan.org>

=item *

Sean M Burke <sburke@cpan.org>

=item *

Gisle Aas <gisle@ActiveState.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Nigel Metheringham, 2002-2005 Sean M Burke, 1999-2002 Gisle Aas.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Base class for HTML formatters",
  "AUTHOR" => "Nigel Metheringham <nigelm\@cpan.org>, Sean M Burke <sburke\@cpan.org>, Gisle Aas <gisle\@ActiveState.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "HTML-Formatter",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "HTML::Formatter",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Encode" => 0,
    "Font::Metrics::Courier" => 0,
    "Font::Metrics::CourierBold" => 0,
    "Font::Metrics::CourierBoldOblique" => 0,
    "Font::Metrics::CourierOblique" => 0,
    "Font::Metrics::Helvetica" => 0,
    "Font::Metrics::HelveticaBold" => 0,
    "Font::Metrics::HelveticaBoldOblique" => 0,
    "Font::Metrics::HelveticaOblique" => 0,
    "Font::Metrics::TimesBold" => 0,
    "Font::Metrics::TimesBoldItalic" => 0,
    "Font::Metrics::TimesItalic" => 0,
    "Font::Metrics::TimesRoman" => 0,
    "HTML::Element" => "3.15",
    "HTML::TreeBuilder" => 0,
    "IO::File" => 0,
    "base" => 0,
    "integer" => 0,
    "parent" => 0,
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Slurper" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96",
    "Test::Warnings" => 0,
    "blib" => "1.01",
    "bytes" => 0,
    "lib" => 0
  },
  "VERSION" => "2.15",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::Dumper" => 0,
  "Encode" => 0,
  "File::Slurper" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Font::Metrics::Courier" => 0,
  "Font::Metrics::CourierBold" => 0,
  "Font::Metrics::CourierBoldOblique" => 0,
  "Font::Metrics::CourierOblique" => 0,
  "Font::Metrics::Helvetica" => 0,
  "Font::Metrics::HelveticaBold" => 0,
  "Font::Metrics::HelveticaBoldOblique" => 0,
  "Font::Metrics::HelveticaOblique" => 0,
  "Font::Metrics::TimesBold" => 0,
  "Font::Metrics::TimesBoldItalic" => 0,
  "Font::Metrics::TimesItalic" => 0,
  "Font::Metrics::TimesRoman" => 0,
  "HTML::Element" => "3.15",
  "HTML::TreeBuilder" => 0,
  "IO::File" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Test::More" => "0.96",
  "Test::Warnings" => 0,
  "base" => 0,
  "blib" => "1.01",
  "bytes" => 0,
  "integer" => 0,
  "lib" => 0,
  "parent" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
